### DISCUSSION TITLE
Change svg image based on entity state

### DISCUSSION THREAD

**Discussion by m90att** [ExperienceLovelace/ha-floorplan#244] [Q&A] [answered]
Hi,

I have a problem that I've googled, searched the community and other discussions for pointers but ultimately, I've found nothing that really fits my situation.  Nor am I experienced enough to troubleshoot beyond basic levels and I'm now beyond confused, so I'm reaching out for some expert advice - I'm sure this is straightforward - I just can't see it.

I've got a floorplan [background] svg that I want to change depending on the time of day.  I've borrowed an example of a time of day sensor which is working well...

```
   timeofday:
      value_template: >-
        {% set time = states.sensor.time.state %}
        {% set elevation = states.sun.sun.attributes.elevation %}
        {% set night_from = '22:00' %}
        {% if -2 < elevation < 4 %}
          {{ 'sunrise' if time < '12:00' else 'sunset' }}
        {% elif time < '06:00' or time > night_from %} night
        {% elif time < '09:00' %} morning
        {% elif time < '19:00' %} day
        {% elif time < night_from %} evening
        {% else %} night
        {% endif %}
```

My yaml image config is:

```
      config:
          image: /local/floorplan/five/floorplan_night.svg
          stylesheet: /local/floorplan/five/home.css
          console_log_level: debug
```

I currently have 3 svgs created, 'day', 'evening', 'night' and I want the night one to be the default.

In inkscape the following holds the background...

![image](https://user-images.githubusercontent.com/56609577/194325497-76b2c71a-3a74-4b2a-ab21-7e0116fb37a8.png)

![image](https://user-images.githubusercontent.com/56609577/194325619-2210416f-2122-468d-82e9-ef64c549acd7.png)

The code I've added to the yaml rules is:

```
            - element: sensor.timeofday
              entity: sensor.timeofday
              tap_action: false
              hover_action: false
              state_action:
                - action: call-service
                  service: floorplan.image_set
                  service_data: |
                      >
                      switch (entity) {
                        case "day":
                          return "/local/floorplan/background/floorplan_day.svg;

                        case "evening":
                          return "/local/floorplan/background/floorplan_evening.svg";

                        default:
                          return "/local/floorplan/background/floorplan_night.svg";

                      }
```

but nothing happens, it always shows floorplan_night.  What am I doing wrong?!

Thank you v much,

Matt

**Comment by exetico** [comment, marked as answer]
Hi

It's great to see people trying. That's the only way to make progress in a new space.

TL;DR: Use `entity.state` in your switch-statement, and you're golden.

**Answer**:
I've created the sensor so the `day`, `evening` and more works as expected.
![image](https://user-images.githubusercontent.com/3549445/194353236-66049c5d-2fec-46b1-b916-70b97a7599ca.png)

My YAML looks like this:
```YAML
element: area.office
entity: sensor.timeofday
tap_action: false
hover_action: false
state_action:
  - action: call-service
    service: floorplan.image_set
    service_data: |
        >
        console.log("Entity Info - Debug info:", entity);
        switch (entity.state) {
          case "day":
            return "/local/floorplan/__test/fp-test-day.svg";
          case "evening":
            return "/local/floorplan/__test/fp-test-evening.svg";
          default:
            return "/local/floorplan/__test/fp-test-default.svg";
        }
```

Please take a look at the `console.log`-line.

Open the Inspect window in your browser, navigate to the console-window, and take a short time to look at the entity-object:
![image](https://user-images.githubusercontent.com/3549445/194354619-0647fffc-b723-404f-9e50-c2cb3aa3b581.png)

Here you'll see that the "entity" contains more than a single value. It's an object, with multiple properties.
Therefore you'll have to grap `entity.state`, instead of just using `entitiy` in the `switch()`.

You _could_ also check if `entity.state` is even defined, before trying to access it... but I guess that's story for another JavaScript DIY-day.

Here's my result, with the test-`day` svg file:
![image](https://user-images.githubusercontent.com/3549445/194354984-ac98180b-538a-4cc6-a0cf-4172930f7272.png)

And here's the result with test-`evening` file:
![image](https://user-images.githubusercontent.com/3549445/194355089-dfd761ee-bc12-4277-864f-8846fae93b72.png)

_Bonus info_:
You'll have to define the "property" for almost every ha-floorplan function, where you're able to use the `entity` object.

Here's a example with the new `dataset_set` function.

The `dataset_set` function can be useful if you're trying to style a specific element, where you **don't** want the class-name to contain the state of the entity - which has been pretty common for many floorplans (like my personal floorplan).
![image](https://user-images.githubusercontent.com/3549445/194360033-d28861cc-10c8-473f-a4c1-eb2a7b00e726.png)

```yaml
element: area.office
entity: sensor.timeofday
tap_action: false
hover_action: false
state_action:
  - action: call-service
    service: floorplan.dataset_set
    service_data: 
      key: timeofday
      value: ${entity.state}
```

Result:
![image](https://user-images.githubusercontent.com/3549445/194356993-e92337f1-9dae-440f-ae1d-9410c70bd2ed.png)

Here's even more about Javascript Objects, if you're ready for a ride: https://www.w3schools.com/js/js_objects.asp

**Reply by m90att** [reply]
Thank you so much @exetico !

I love answers like this, with explanations and examples.  It really helped to understand it.  It almost makes my two days of frustration worth it...almost!!

I tried copying and pasting but I think I did it incorrectly as I was getting errors, so I typed it out fresh and, hey presto, it works.

You're a star!!

**Reply by exetico** [reply]
I'm glad that it worked as expected!

Thank you for the kind words, too :smiley: .
